File: cart-service/internal/migration/v003_add_schema_validation.go
package migration

import (
	"context"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
)

// V003AddSchemaValidation adds MongoDB JSON schema validation to the carts collection
type V003AddSchemaValidation struct{}
func NewV003AddSchemaValidation() *V003AddSchemaValidation { return &V003AddSchemaValidation{} }
func (m *V003AddSchemaValidation) ID() string     { return "V003_AddSchemaValidation" }
func (m *V003AddSchemaValidation) Order() string  { return "003" }
func (m *V003AddSchemaValidation) Author() string { return "emart-db-team" }

func (m *V003AddSchemaValidation) Execute(ctx context.Context, db *mongo.Database) error {
	jsonSchema := bson.M{
		"$jsonSchema": bson.M{
			"bsonType": "object",
			"required": []string{"user_id", "items", "total_items", "total_price"},
			"properties": bson.M{
				"user_id":    bson.M{"bsonType": "string", "description": "User ID from login service"},
				"items":      bson.M{"bsonType": "array", "description": "Array of cart items"},
				"total_items": bson.M{"bsonType": "int", "minimum": 0},
				"total_price": bson.M{"bsonType": "double", "minimum": 0},
				"schema_version": bson.M{"bsonType": "int", "minimum": 1},
			},
		},
	}

	cmd := bson.D{
		{Key: "collMod", Value: "carts"},
		{Key: "validator", Value: jsonSchema},
		{Key: "validationLevel", Value: "moderate"},
		{Key: "validationAction", Value: "error"},
	}

	return db.RunCommand(ctx, cmd).Err()
}

func (m *V003AddSchemaValidation) Rollback(ctx context.Context, db *mongo.Database) error {
	cmd := bson.D{
		{Key: "collMod", Value: "carts"},
		{Key: "validator", Value: bson.M{}},
		{Key: "validationLevel", Value: "off"},
	}
	return db.RunCommand(ctx, cmd).Err()
}

File: scripts/db/V003_AddSchemaValidation.js
// ============================================================
// V003 - Add JSON Schema Validation to carts collection
// DB Team standalone script
// Usage: mongosh "mongodb://user:pass@host:27017" --file V003_AddSchemaValidation.js
// ============================================================

use('cart');

print('=== V003_AddSchemaValidation START ===');

// Check if already executed
var existing = db.mongockChangeLog.findOne({
    changeId: 'V003_AddSchemaValidation',
    state: 'EXECUTED'
});

if (existing) {
    print('  ⏭ V003 already executed at ' + existing.executedAt + ' - skipping');
    quit(0);
}

// Apply JSON Schema validation to the carts collection
db.runCommand({
    collMod: 'carts',
    validator: {
        $jsonSchema: {
            bsonType: 'object',
            required: ['user_id', 'items', 'total_items', 'total_price'],
            properties: {
                user_id: {
                    bsonType: 'string',
                    description: 'Required. User ID from Login service JWT. Must be a string.'
                },
                items: {
                    bsonType: 'array',
                    description: 'Required. Array of cart items.',
                    items: {
                        bsonType: 'object',
                        required: ['item_id', 'product_id', 'product_name', 'category', 'price', 'quantity'],
                        properties: {
                            item_id:      { bsonType: 'string' },
                            product_id:   { bsonType: 'string' },
                            product_name: { bsonType: 'string' },
                            category: {
                                bsonType: 'string',
                                // Only these 3 categories are valid - matches cart service handler
                                enum: ['books', 'courses', 'software'],
                                description: 'Must be one of: books, courses, software'
                            },
                            price: {
                                bsonType: 'double',
                                minimum: 0,
                                description: 'Price must be >= 0'
                            },
                            quantity: {
                                bsonType: 'int',
                                minimum: 1,
                                maximum: 100
                            }
                        }
                    }
                },
                total_items: {
                    bsonType: 'int',
                    minimum: 0
                },
                total_price: {
                    bsonType: 'double',
                    minimum: 0
                },
                schema_version: {
                    bsonType: 'int',
                    minimum: 1
                }
            }
        }
    },
    validationLevel:  'moderate', // applies to inserts & updates; existing docs unaffected
    validationAction: 'error'     // reject invalid documents
});

print('  ✓ JSON Schema validation applied to carts collection');
print('  ✓ Valid categories: books, courses, software');
print('  ✓ Required fields: user_id, items, total_items, total_price');
print('  ✓ validationLevel: moderate | validationAction: error');

// Record in Mongock change log
db.mongockChangeLog.updateOne(
    { changeId: 'V003_AddSchemaValidation' },
    {
        $set: {
            changeId:   'V003_AddSchemaValidation',
            author:     'emart-db-team',
            state:      'EXECUTED',
            executedAt: new Date(),
            order:      '003',
            source:     'manual-script'
        }
    },
    { upsert: true }
);

print('=== V003_AddSchemaValidation DONE ===');
